source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -193,7 +216,8 @@ absl::Status ProcessorState::handleHeadersResponse(const HeadersResponse& respon
         // let the doData callback handle body chunks until the end is reached.
         clearWatermark();
         return absl::OkStatus();
-      } else if (body_mode_ == ProcessingMode::STREAMED) {
+      } else if (body_mode_ == ProcessingMode::STREAMED ||
+                 body_mode_ == ProcessingMode::FULL_DUPLEX_STREAMED) {
```

Comment:
In FULL_DUPLEX_STREAMED mode, we send the body and trailer as they arrive. i.e., we don't hold them while we are waiting for the header response.
 
 QQ: In which cases, we will need to send buffered body in handleHeadersResponse function here?

Reply:
For both STREAMED and FULL_DUPLEX_STREAMED mode, if flow control causes some data are stored in Envoy buffer where waiting for header response, this is the chance to send them.  If there is no data buffered, then with this check https://github.com/envoyproxy/envoy/blob/72b75074a0ee089ad81f68ae011e31f14c2936fe/source/extensions/filters/http/ext_proc/processor_state.cc#L87, this function call will be skipped.

Reply:
Flow control should not add the data to this particular buffer. Can you clarify in which cases specifically or do you have the test cases that FULL_DUPLEX_STREAMED will hit this case, that you can provide?

Reply:
That's a good question.

Let's first clarify why this is needed for STREAMED mode.

IIUC, in STREAMED mode, if clients is sending huge amount of data, which is buffered in the queue, then flow control triggered, and no more data is able to be put into the queue.  However, there could be some data be buffered somewhere else in the Envoy, like previous filter in the filter chain as described here:
https://github.com/envoyproxy/envoy/blob/72b75074a0ee089ad81f68ae011e31f14c2936fe/envoy/http/filter.h#L551

And this is the reason why for STREAMED mode, there is handling here, which basically sends the buffered data as described above .  If there is no buffer outside of the input queue for STREAMED case, then there is no need for the handling here even for STREAMED mode, right?

One question I would like to ask you is that, with the new flow control you added recently, when flow control gets triggered and the side stream starts to push back the downstream, does Envoy buffer some data somewhere, or, there is no Envoy buffering at all?

Reply:
First, if you take a closer look, my comment has always been around `FULL_DUPLEX_STREAMED `. I know why STREAM_MODE has it and the STREAM_MODE discussion is not relevant. 

Second, to your question. The new flow control will not add the data to this particular buffer.

As i mentioned above, let's check if you have test cases that FULL_DUPLEX_STREAMED will hit this line. That should be easy to check with some logging and can help to guide next steps.

Reply:
Thanks for the comments!
This newly added test: TEST_P(ExtProcIntegrationTest, ServerSendBodyRespWithouRecvEntireBodyDuplexStreamed) will hit this function call here: sendBufferedDataInStreamedMode().  With Envoy gRPC test case, this test triggers the flow control as shown in below message in the log:
[2024-11-02 18:07:00.072][63][debug][connection] [source/common/network/connection_impl.cc:594] [Tags: "ConnectionId":"13"] onAboveWriteBufferHighWatermark
[2024-11-02 18:07:00.073][63][debug][http] [source/common/http/conn_manager_impl.cc:1935] [Tags: "ConnectionId":"12","StreamId":"4233291703796802150"] Read-disabling downstream stream due to filter callbacks.

However, inside the function sendBufferedDataInStreamedMode(),   the hasBufferedData() check never returns true, i.e, the data is not buffered at all.

With that, let me skip calling sendBufferedDataInStreamedMode() for MXN mode here.

Reply:
Thanks for confirming! Yep this matches my suspicion at the beginning of this thread that FULL_DUPLEX_STREAMED doesn't have the buffered body here.

Update here LGTM